File: go/quantum_state.go
package quantum

import (
    "math"
    "math/rand"
)

type QuantumState struct {
    state     []float64
    numQubits int
}

func NewQuantumState(numQubits int) *QuantumState {
    size := 1 << numQubits
    state := make([]float64, size)
    state[0] = 1.0
    return &QuantumState{
        state:     state,
        numQubits: numQubits,
    }
}

func (q *QuantumState) ApplyGate(gate [][]float64, target int) {
    size := 1 << q.numQubits
    result := make([]float64, size)
    gateSize := len(gate)

    for i := 0; i < size; i++ {
        sum := 0.0
        for j := 0; j < gateSize; j++ {
            mask := 1 << target
            bitI := (i & mask) >> target
            bitJ := (j & mask) >> target

            index := (i & ^mask) | (bitJ << target)
            sum += gate[j][bitI] * q.state[index]
        }
        result[i] = sum
    }

    q.state = result
}

func (q *QuantumState) Measure() int {
    probabilities := make([]float64, len(q.state))
    total := 0.0

    for i, val := range q.state {
        probabilities[i] = math.Pow(val, 2)
        total += probabilities[i]
    }

    for i := range probabilities {
        probabilities[i] /= total
    }

    r := rand.Float64()
    cumulative := 0.0

    for i, prob := range probabilities {
        cumulative += prob
        if r <= cumulative {
            newState := make([]float64, len(q.state))
            newState[i] = 1.0
            q.state = newState
            return i
        }
    }

    return 0
}

File: rust/quantum_state.rs
use nalgebra::{DMatrix, DVector};
use rand::prelude::*;
use std::f64::consts::PI;

struct QuantumState {
    state: DVector<f64>,
    num_qubits: usize,
}

impl QuantumState {
    fn new(num_qubits: usize) -> Self {
        let size = 1 << num_qubits;
        let mut state = DVector::zeros(size);
        state[0] = 1.0;
        QuantumState { state, num_qubits }
    }

    fn apply_gate(&mut self, gate: &DMatrix<f64>, target: usize) {
        let mut result = DVector::zeros(1 << self.num_qubits);
        let gate_size = gate.nrows();
        
        for i in 0..(1 << self.num_qubits) {
            let mut sum = 0.0;
            for j in 0..gate_size {
                let mask = 1 << target;
                let bit_i = (i & mask) >> target;
                let bit_j = (j & mask) >> target;
                
                let index = (i & !mask) | (bit_j << target);
                sum += gate[(j, bit_i)] * self.state[index];
            }
            result[i] = sum;
        }
        
        self.state = result;
    }

    fn measure(&mut self) -> usize {
        let mut rng = thread_rng();
        let probabilities: Vec<f64> = self.state.iter().map(|&x| x.powi(2)).collect();
        let total: f64 = probabilities.iter().sum();
        let probabilities: Vec<f64> = probabilities.iter().map(|&x| x / total).collect();
        
        let mut cumulative = 0.0;
        let r: f64 = rng.gen();
        
        for (i, &prob) in probabilities.iter().enumerate() {
            cumulative += prob;
            if r <= cumulative {
                self.state.fill(0.0);
                self.state[i] = 1.0;
                return i;
            }
        }
        
        0
    }
}
